cpp/src/wholegraph_ops/raft_random_gen.cpp

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2019-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <wholememory/wholegraph_op.h>
+#include <cmath>
+#include <wholememory_ops/raft_random.cuh>
```

Comment:
several things here:
1. please open an issue with RAFT to include host functionality for their random generators, we should not add duplicate code with RAFT here
2. this file cannot be compiled by a non-CUDA compiler, so please use the `.cu` extension
3. Is there a reason why you cannot expose the full RNG state on your API ? the RNG random state for RAFT (`raft::random::RngState`) is opaque and should be used for APIs instead of using just the seed, since otherwise, you may not get the desired behavior at some point in the future. You should avoid construction the generator based on seed and subsequence if possible, and rather construct it based on the state directly (currently, in RAFT, you have to first create the templated `DeviceState`).

Reply:
I have opened an issue with RAFT to include host functionality for their random generators.
For the third problem,  these two functions are packed into Python functions and aimed to test the correctness of Python GPU weighted and unweighted sample ops.  For precise control of the random value generated, we needed the subsequence in the Python test code.  raft::random::RngState has a public member variable of base_subsequence. But it seems it cannot be used in Python code.

Reply:
I'm not sure I understand why `raft::random::RngState` cannot be used in python code. Do you get a specific error when trying to create the bindings?
It looks to me like the RngState in particular is already exposed in cython in pylibraft: https://github.com/rapidsai/raft/blob/branch-23.08/python/pylibraft/pylibraft/random/cpp/rng_state.pxd

Reply:
> We need the internal subsequence of the random generator.

This is exactly my point, you should not rely on anything internal / implementation detail of RAFT. One thing you can add to the issue for RAFT is that the host functionality should emulate device functionality when given the same seeds/base-subsequences.
In order to just move ahead here, we can ignore this for now, but please open an issue on this repo as well to remove this functionality in favor of the RAFT implementation.

> Maybe, is it better for Pylibraft to provide this binding and expose it as a Python interface?

As far as I understand the file linked above, you should be able to do `from pylibraft.random.cpp.rng_state cimport RngState` in your cython files. So my question was: why are you not doing this ?
Additionally, you can also add to the issue in RAFT that we should expose functionality around `RngState` to python, although I'm not sure exactly what you would need (more than the cython definition of the C++ struct).

Reply:
Finally, can you also explain why you have to generate random numbers on the CPU at all? Can we not simply use the same random numbers generated on the GPU for both the CPU-based and GPU-based algorithms? What am I missing here?

Reply:
We need to test the correctness of GPU weighted and unweighted random sampling ops. There are random values to generate in these two ops on GPU.  For the reference of CPU implementation of these two ops, we need to generate the same random values as the GPU. So we need to precisely control what random values to generate, which needs the internal subsequence of raft random generator.  If we don't use the same random numbers generated on GPU and CPU, the results of weighted and unweighted random sampling ops on the GPU are not the same as that of CPU.

Reply:
> This is exactly my point, you should not rely on anything internal / implementation detail of RAFT. One thing you can add to the issue for RAFT is that the host functionality should emulate device functionality when given the same seeds/base-subsequences. In order to just move ahead here, we can ignore this for now, but please open an issue on this repo as well to remove this functionality in favor of the RAFT implementation.

"the host functionality should emulate device functionality when given the same seeds/base-subsequences"  is our need. I have added this issue for RAFT. And, I will open an issue on this repo to remove this functionality when RAFT implementation is done. 

> As far as I understand the file linked above, you should be able to do `from pylibraft.random.cpp.rng_state cimport RngState` in your cython files. So my question was: why are you not doing this ? Additionally, you can also add to the issue in RAFT that we should expose functionality around `RngState` to python, although I'm not sure exactly what you would need (more than the cython definition of the C++ struct).

I also added to the issue in RAFT.

Reply:
> So we need to precisely control what random values to generate, which needs the internal subsequence of raft random generator.

But why not simply generate the values once, then use them for both CPU and GPU implementations?
Or use the same state in both cases?
Do you have to run the CPU tests on a machine without GPU?

If you would like to check correctness of the generator itself, this should be done by RAFT (or you should not use their generator if it cannot be trusted).

Reply:
I've looked a bit more into your implementation for unweighted sampling. There are a few other flaws for which I'll open issues.

Overall, I would highly recommend to use the `RngState` on APIs and `DeviceState` within kernels. Once we add host functionality to RAFT, you can also construct the `DeviceState` on the host and construct the generator the same way as in the kernel.

Reply:
I'll open issues here for this, so for now, feel free to merge this PR